presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -437,6 +443,36 @@ private Estimate calculateNullsCount(Map<String, PartitionStatistics> statistics
                 });
     }
 
+    private Estimate calculateDistinctValuesCountForPartitioningKey(HiveColumnHandle hiveColumnHandle, List<HivePartition> partitions)
+    {
+        return new Estimate(partitions.stream()
+                .map(partition -> partition.getKeys().get(hiveColumnHandle))
+                .collect(toSet())
+                .size());
+    }
+
+    private Estimate calculateNullsCountForPartitioningKey(HiveColumnHandle hiveColumnHandle, List<HivePartition> hivePartitions, Map<String, PartitionStatistics> partitionStatisticsMap)
+    {
+        List<Long> knownPartitionRowCounts = partitionStatisticsMap.values().stream()
```

Comment:
Instead of materializing the values and computing the sum and the size to get the average, just do:

```java
private Estimate calculateNullsCountForPartitioningKey(HiveColumnHandle hiveColumnHandle, List<HivePartition> partitions, Map<String, PartitionStatistics> statistics)
{
    OptionalDouble rowsPerPartition = statistics.values().stream()
            .map(PartitionStatistics::getRowCount)
            .filter(OptionalLong::isPresent)
            .mapToLong(OptionalLong::getAsLong)
            .average();

    if (!rowsPerPartition.isPresent()) {
        return Estimate.unknownValue();
    }

    return new Estimate(partitions.stream()
            .filter(partition -> partition.getKeys().get(hiveColumnHandle).isNull())
            .map(HivePartition::getPartitionId)
            .mapToLong(partitionId -> statistics.get(partitionId).getRowCount().orElse((long) rowsPerPartition.getAsDouble()))
            .sum());
}
```